resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,265 @@
+
+- **DOWN**(`T Function<...>(...)`, `S Function<...>(...)`) = `bottom` otherwise
+
+
+- **DOWN**(`T1`, `T2`) = `T1` if `T1` <: `T2`
+- **DOWN**(`T1`, `T2`) = `T2` if `T2` <: `T1`
+- **DOWN**(`T1`, `T2`) = `bottom` otherwise
+
+
+## Issues and Interesting examples
+
+### Type variable bounds
+
+The definition of upper bound for type variables does not guarantee termination.
+Counterexample:
+
+```dart
+void foo<T extends List<S>, S extends List<T>>() {
+  T x;
+  S y;
+  var a = (x == y) ? x : y;
+}
+```
+
+It should be changed to close the bound with respect to all of the type
+variables declared in the same scope, using the greatest closure definition.
+
+### Generic functions
+
+The CFE currently implements upper bounds for generic functions incorrectly. Example:
+
+```dart
+typedef G0 = T Function<T>(T x);
+typedef G1 = T Function<T>(T x);
+void main() {
+   {
+    G0 x;
+    G1 y;
+    // Analyzer: T Function<T>(T)
+    // CFE: bottom -> Object
+    var a = (x == y) ? x : y;
+
+    }
+}
+```
+
+Both the CFE and the analyzer currently implement lower bounds for generic
+functions incorrectly.  Example:
+
+```dart
+typedef G0 = T Function<T>(T x);
+typedef G1 = T Function<T>(T x);
+void main() {
+  {
+    void Function(G0) x;
+    void Function(G1) y;
+    int z;
+    // Analyzer: void Function(Never Function(Object))
+    // CFE:      void Function(bottom-type Function(Object))
+    var a = (x == y) ? x : y;
+  }
+}
+```
+
+## Asymmetry
+
+The current algorithm is asymmetric.  There is an equivalence class of top
+types, and we correctly choose a canonical representative for bare top types
+using the **MORETOP** predicate.  However, when two different top types are
+embedded in two mutual subtypes, we don't correctly choose a canonical
+representative.
+
+```dart
+import 'dart:async';
+
+void main () {
+  List<FutureOr<Object>> x;
+  List<dynamic> y;
+  String s;
+  // List<dynamic>
+  var a = (x == y) ? x : y;
+  // List<FutureOr<Object>>
+  var b = (x == y) ? y : x;
+```
+
+The best solution for this is probably to normalize the types.  This is fairly
+straightforward: we just normalize `FutureOr<T>` to the normal form of `T` when
+`T` is a top type.  We can then inductively apply this across the rest of the
+types.  Then, whenever we have mutual subtypes, we just return the normal form.
+This would be breaking, albeit hopefully only in a minor way.
+
+An alternative would be to try to define an ordering on mutual subtypes.  This
+can probably be done, but is a bit ugly.  For example, consider `Map<dynamic,
+FutureOr<dynamic>>` vs `Map<FutureOr<dynamic>, dynamic>`.  The obvious way to
+proceed is to define the total order by defining a traversal order on types, and
+then defining the ordering lexicographically.  That is, saying that `T` is
+greater than `S` if the first pair of top types encountered in the traversal
+that are not identical are `T0` and `S0` respectively, and **MORETOP**(`T0`,
+`S0`).
```

Comment:
In https://dart-review.googlesource.com/c/sdk/+/84908 I proposed to refine the regular subtype relationship such that it would take the top type ordering into account. I believe this would amount to copying the subtype ruleset (for each `<:` rule we'd get a copy which uses `<::` everywhere), and then we'd have the MORETOP rules rather than the existing `<:` rules about top types, and there would be a `T <: FutureOr^k<Object>`-for-all-k rule to connect the top types to other types.

So I don't think we'd need anything like lexicographic ordering. For instance, the least upper bound of `Map<dynamic, FutureOr<dynamic>>` and `Map<FutureOr<dynamic>, dynamic>` would be `Map<dynamic, dynamic>`.

Reply:
Could we define the more precise relation first, then define the current subtype relation as a collapse of that other relation where top types are considered equivalent?

Reply:
@eernstg I'm not sure what you're proposing.  What, concretely, do you want to return for this algorithm in the case that I have `T1 <: T2` and `T2 <: T1` for arbitrary `T1, T2`?

@lrhn I don't follow what you're proposing here.

Reply:
No, I don't care about resource economy here at all.  I just don't know what you are concretely proposing.  Let me try to interpret what you are saying a bit.  You are proposing a relation `<::` which you describe as "finer grained", and say that you would just compute that, instead of the subtype check.  How is this relation different from my suggestion to impose a total order on mutual subtypes?  I think it is different, since you say above that 

> the least upper bound of `Map<dynamic, FutureOr<dynamic>>` and `Map<FutureOr<dynamic>, dynamic>` would be `Map<dynamic, dynamic>`.

This implies that if `T` is the first type, and `S` is the second type there, that neither `S <:: T` nor `T <:: S`.  So maybe you mean that instead of getting a total order by lexicographically sorting, you'll get a partial order by pointwise comparison, and then the idea is that if the two types are mutual subtypes but are not related in the partial order, then you'll end up decomposing them recursively in an appropriate way.  That may be true for the top types.  This is connected to a different idea I had.  I don't think it quite works for `FutureOr<Null>` vs `Future<Null>` though - those are presumably not ordered by your partial order, but they are mutual subtypes, and they will not be recursively decomposed.

I'm not opposed to trying to approach it this way, but it feels very complicated to me, since you need to specify a variant of subtyping with some very particular properties.  Specifically, you need:
  - `T <:: S` => `T <: S`
  - `T <: S` and `S <: T` and `S != T` => not (`S <:: T` or `T <:: S`)

This seems to require a lot of complex reasoning about all of the ways that two types could be subtypes.

Reply:
Yes, the idea I had in mind is to specify a relation `<::` on types such that `T <:: S => T <: S` (that's what I mean by "more fine grained"), and such that it is a total order on top types and bottom types. It would be a complete copy of the regular subtype rules, except that we would replace the rules that make top types equivalent (`T <: void`, `T <: dynamic`, `T <: Object?`) by a single rule  connecting all other types to the top type chain (maybe `T <: FutureOr^\infty<Object?>`, whenever `T` is not a top type).

We would then have `LUB(FutureOr<dynamic>, dynamic) == LUB(dynamic, FutureOr<dynamic>) == dynamic`, so the ordinary "same interface type" rule for LUB would give us `LUB(Map<dynamic, FutureOr<dynamic>>, Map<FutureOr<dynamic>, dynamic>) == Map<dynamic, dynamic>`. The recursive application of LUB comes from the LUB rules (no changes needed there), so the choice to recursion is not a property of `<::` at all.

It is certainly possible that a model where we consider equivalence classes of types (based on mutual subtyping) and then specify some LUB that uses `<:` in some cases and `MORETOP`/`MOREBOTTOM` in other cases would turn out to be equivalent, but I think that would be more complex than having a uniformly defined `<::` and using that in the LUB rules.

Also, `<::` could also be used to determine whether there are any conflicts when we compute the interface of a class.

Reply:
> (maybe `T <: FutureOr^\infty<Object?>`, whenever `T` is not a top type).

This doesn't feel well-defined to me, I think we'd need to find a different way to specify it.  FutureOr^\infty is not a term in the language - if it were, then we'd lose well-founded induction.

It's possible we can take something like this as a route, but it seems complex to me.  In any case, it's not part of this PR.

Reply:
Added a brief note on this option.

Reply:
I guess my decidedly underspecified idea was to start out defining a relation where `... FutureOr<void> <:: Object <:: dynamic <:: void` etc, and then defining `<:` as `<::` modulo the equivalence `Object` == `dynamic` == `void` and `FutureOr<T>` == `T` iff `Future<T> <: `T`, etc.

Maybe that "relation module an equivalence" isn't actually simple to do, at least we'd have to prove that all elements in each equivalence class are related in the same way to all other elements. I just hoped that it might be easier than defining two relations in their entirety, where part of them are just copies of the same rules.

Reply:
There might be meta-syntactic tricks we could play to make this definition relatively compact.  I think what I would probably try to do if I were to try to formalize this approach is to index the relation, and have the fine grained rules only present for one of the indices.  By which I mean, define the relation as

`SUBTYPE(T, S, k)` where `T` and `S` are types, and `k` ranges over `<:` and `<::`.

Now most of your rules are parametric over `k`.  So you have:

`SUBTYPE(T?, S, k)` iff `SUBTYPE(T, S, k)` and `SUBTYPE(Null, S, k)`

But some of the top rules are specialized on `k`.  So you have something like:

`SUBTYPE(dynamic, void, k)` as an axiom
`SUBTYPE(void, dynamic, <:)` as an axiom

I would hypothesize that if you set it up like that, then you basically get the theorem that `SUBTYPE(T, S, <::) => SUBTYPE(T, S, <:)` for free: every derivation of the former has a valid derivation in the latter, since there are no rules which only apply to `<::`.